### PR TITLE
m_lights

### PR DESCRIPTION
**Pull request by Kelebek1**

### CODE REVIEW

File: include/macros.h [hunk 1]
```diff
@@ -18,4 +18,8 @@
 
 #define ABS(x) (((x) >= 0) ? (x): -(x))
 
+#define CLAMP(x, min, max) ((x) < (min) ? (min) : (x) > (max) ? (max) : (x))
+#define CLAMP_MAX(x, max) ((x) > (max) ? (max) : (x))
+#define CLAMP_MIN(x, min) ((x) < (min) ? (min) : (x))
```
Comment (AngheloAlf) [inline include/macros.h:23]: Why? They don't seem to be used

File: lib/ultralib/include/PR/gbi.h [hunk 2]
```diff
@@ -1353,11 +1353,11 @@ typedef struct {
...
 typedef struct {
     unsigned char col[3];
-    unsigned char unk3;
+    unsigned char kc;
     unsigned char colc[3];
-    unsigned char unk7;
+    unsigned char kl;
     short pos[3];
-    unsigned char unkE;
+    unsigned char kq;
```
Comment (AngheloAlf) [inline lib/ultralib/include/PR/gbi.h:1360]: What does those names mean?

File: src/code/m_lights.c [hunk 3]
```diff
@@ -0,0 +1,366 @@
...
+    lights->lights.a.l.col[1] = lights->lights.a.l.colc[1] = g;
+    lights->lights.a.l.col[2] = lights->lights.a.l.colc[2] = b;
+    lights->diffuse_count = 0;
+}
+
+void LightsN_disp(LightsN* lights, GraphicsContext* gfxCtx) {
+    Light* light;
+    s32 i;
+    Gfx* opa_gfx;
+    Gfx* xlu_gfx;
```
Comment (AngheloAlf) [inline src/code/m_lights.c:58]:
```suggestion
    Gfx* opaGfx;
    Gfx* xluGfx;
```

File: src/code/m_lights.c [hunk 4]
```diff
@@ -0,0 +1,366 @@
...
+    f32 zdiff;
+    f32 pointd;
+    f32 rad;
+    Light* light;
+
+    if ((point != NULL) && (lightInfo->point.radius > 0)) {
+        xdiff = lightInfo->point.x - point->x;
+        ydiff = lightInfo->point.y - point->y;
+        zdiff = lightInfo->point.z - point->z;
+        pointd = (xdiff * xdiff) + (ydiff * ydiff) + (zdiff * zdiff);
```
Comment (AngheloAlf) [inline src/code/m_lights.c:107]: You could use the `SQ` macro here

File: src/code/m_lights.c [hunk 5]
```diff
@@ -0,0 +1,366 @@
...
+
+        rad = lightInfo->point.radius;
+
+        if (pointd < (rad * rad)) {
```
Comment (AngheloAlf) [inline src/code/m_lights.c:111]: Same here about the `SQ` macro

File: src/code/m_lights.c [hunk 6]
```diff
@@ -0,0 +1,366 @@
...
+    }
+    light_list_buf.current++;
+    return lightNode;
+}
+
+void Light_list_buf_delete(LightNode* lightNode) {
+    if (lightNode != NULL) {
+        light_list_buf.current--;
+        lightNode->info = NULL;
+        light_list_buf.idx = (lightNode - light_list_buf.lights) / (s32)sizeof(LightNode);
```
Comment (AngheloAlf) [inline src/code/m_lights.c:228]: Is the `s32` cast required here?

File: src/code/m_lights.c [hunk 7]
```diff
@@ -0,0 +1,366 @@
...
+#include "gfx.h"
+#include "macros.h"
+#include "sys_matrix.h"
+#include "overlays/gamestates/ovl_play/m_play.h"
+#include "6E0F50.h"
+
+extern Gfx D_400AA00[];
+extern Gfx D_400AA40[];
+
+static LightBuffer light_list_buf;
```
Comment (AngheloAlf) [inline src/code/m_lights.c:12]: why the `static`?

File: include/m_lights.h [hunk 8]
```diff
@@ -4,42 +4,90 @@
...
-// void func_8009BB8C_jp();
-// void func_8009BBEC_jp();
-// void func_8009BC48_jp();
-// void func_8009BD90_jp();
+    /* 0x0 */ LightNode* list;
+    /* 0x4 */ u8 ambientColor[3];
+    /* 0x7 */ u8 fogColor[3];
+    /* 0xA */ s16 fogNear;
+    /* 0xC */ s16 fogFar; 
+} LightContext; // size = 0x10
```
Comment (AngheloAlf) [inline include/m_lights.h:61]: Looking at the gcn version this struct should be renamed to `Global_light` instead

File: src/code/m_lights.c [hunk 9]
```diff
@@ -0,0 +1,366 @@
...
+
+void point_data_set(Lights* lights, s16 x, s16 y, s16 z, u8 r, u8 g, u8 b, s16 radius, s32 type) {
+    lights->type = type;
+    lights->lights.point.x = x;
+    lights->lights.point.y = y;
+    lights->lights.point.z = z;
+    Light_point_color_set(lights, r, g, b, radius);
+}
+
+void func_8009B23C_jp(Lights* lights, s16 x, s16 y, s16 z, u8 r, u8 g, u8 b, s16 radius) {
```
Comment (AngheloAlf) [inline src/code/m_lights.c:22]: This function seems to be `Light_point_ct`

File: src/code/m_lights.c [hunk 10]
```diff
@@ -0,0 +1,366 @@
...
+
+LightsN* Global_light_read(LightContext* lightCtx, struct GraphicsContext* gfxCtx) {
+    return new_LightsN(gfxCtx, lightCtx->ambientColor[0], lightCtx->ambientColor[1], lightCtx->ambientColor[2]);
+}
+
+void Global_light_list_ct(LightContext* lightCtx) {
+    lightCtx->list = NULL;
+}
+
+void func_8009BB40_jp(LightContext* lightCtx) {
```
Comment (AngheloAlf) [inline src/code/m_lights.c:261]: This seems to be `Global_light_list_dt`

File: src/code/m_lights.c [hunk 11]
```diff
@@ -0,0 +1,366 @@
...
+    lights->type = 1;
+    lights->lights.diffuse.x = x;
+    lights->lights.diffuse.y = y;
+    lights->lights.diffuse.z = z;
+    lights->lights.diffuse.color[0] = r;
+    lights->lights.diffuse.color[1] = g;
+    lights->lights.diffuse.color[2] = b;
+}
+
+void func_8009B394_jp(LightsN* lights, u8 r, u8 g, u8 b) {
```
Comment (AngheloAlf) [inline src/code/m_lights.c:47]: This could be `LightsN_ct`

File: src/code/m_lights.c [hunk 12]
```diff
@@ -0,0 +1,366 @@
...
+    lights->lights.point.y = y;
+    lights->lights.point.z = z;
+    Light_point_color_set(lights, r, g, b, radius);
+}
+
+void func_8009B23C_jp(Lights* lights, s16 x, s16 y, s16 z, u8 r, u8 g, u8 b, s16 radius) {
+    point_data_set(lights, x, y, z, r, g, b, radius, 0);
+}
+
+void func_8009B2A0_jp(Lights* lights, s16 x, s16 y, s16 z, u8 r, u8 g, u8 b, s16 radius) {
```
Comment (AngheloAlf) [inline src/code/m_lights.c:26]: maybe `Light_point2_ct`

File: src/code/m_lights.c [hunk 13]
```diff
@@ -0,0 +1,366 @@
...
+            lightCtx->list = lightNode->next;
+        }
+        if (lightNode->next != NULL) {
+            lightNode->next->prev = lightNode->prev;
+        }
+        Light_list_buf_delete(lightNode);
+    }
+}
+
+LightsN* func_8009BC48_jp(GraphicsContext* gfxCtx, u8 ambient_r, u8 ambient_g, u8 ambient_b, u8 count, u8 light_r,
```
Comment (AngheloAlf) [inline src/code/m_lights.c:297]: `new_Lights` probably

File: src/code/m_lights.c [hunk 14]
```diff
@@ -0,0 +1,366 @@
...
+void Light_list_point_draw(PlayState* play) {
+    LightNode* lightNode = play->lightCtx.list;
+    Gfx* gfx;
+
+    OPEN_DISPS(play->state.gfxCtx);
+    gfx = func_800BD7C0_jp(POLY_XLU_DISP);
+
+    gDPSetAlphaDither(gfx++, G_AD_NOISE);
+    gDPSetColorDither(gfx++, G_CD_MAGICSQ);
+    gSPDisplayList(gfx++, D_400AA00);
```
Comment (AngheloAlf) [inline src/code/m_lights.c:343]: This dlist seems to be `point_light_init_model`, but we don't have a way to name objects' names, so leaving a comment with the name should be fine.

File: src/code/m_lights.c [hunk 15]
```diff
@@ -0,0 +1,366 @@
...
+            f32 rad = (lightParams->point.radius * lightParams->point.radius) * 0.0000026f;
+
+            gDPSetPrimColor(gfx++, 0, 0, lightParams->point.color[0], lightParams->point.color[1],
+                            lightParams->point.color[2], 50);
+
+            Matrix_translate(lightParams->point.x, lightParams->point.y, lightParams->point.z, 0);
+            Matrix_scale(rad, rad, rad, 1);
+
+            gSPMatrix(gfx++, _Matrix_to_Mtx_new(play->state.gfxCtx), G_MTX_NOPUSH | G_MTX_LOAD | G_MTX_MODELVIEW);
+            gSPDisplayList(gfx++, D_400AA40);
```
Comment (AngheloAlf) [inline src/code/m_lights.c:358]: Same here. it seems to be `point_light_model`

File: linker_scripts/jp/symbol_addrs_code.txt [hunk 16]
```diff
@@ -6087,7 +6087,7 @@ B_FLT_80141DF8_jp = 0x80141DF8; // type:f32
 B_FLT_80141DFC_jp = 0x80141DFC; // type:f32
 B_80141E00_jp = 0x80141E00; //
 B_80141E04_jp = 0x80141E04; // type:s16
-B_80141E10_jp = 0x80141E10; //
+light_list_buf = 0x80141E10; //
```
Comment (AngheloAlf) [inline linker_scripts/jp/symbol_addrs_code.txt:6090]:
```suggestion
light_list_buf = 0x80141E10; // size:0x0x188
```

File: linker_scripts/jp/symbol_addrs_code.txt [hunk 17]
```diff
@@ -6087,7 +6087,7 @@ B_FLT_80141DF8_jp = 0x80141DF8; // type:f32
...
 B_80141E18_jp = 0x80141E18; //
 zelda_arena = 0x80141FA0; //
```
Comment (AngheloAlf) [inline linker_scripts/jp/symbol_addrs_code.txt:6092]: remove the fake symbol
```suggestion
light_list_buf = 0x80141E10; //
zelda_arena = 0x80141FA0; //
```

File: lib/ultralib/include/PR/gbi.h (hunk 2 above)
Comment (Kelebek1) [reply lib/ultralib/include/PR/gbi.h:1360]: No idea, it's what they're named in GC: https://github.com/Prakxo/ac-decomp/blob/master/include/libforest/gbi_extensions.h#L335-L342

File: include/macros.h (hunk 1 above)
Comment (Kelebek1) [reply include/macros.h:23]: I added them for testing the kq mis-match, but doesn't seemto hurt to add them anyway?

File: lib/ultralib/include/PR/gbi.h (hunk 2 above)
Comment (Prakxo) [reply lib/ultralib/include/PR/gbi.h:1360]: The names are exposed in the graphics emulator

File: include/macros.h (hunk 1 above)
Comment (AngheloAlf) [reply include/macros.h:23]: The repo already has `MAX` and `MIN` macros which does the same as `CLAMP_MIN` and `CLAMP_MAX` respectively. There has been complains about the inverted names being confusing and there isn't really a need for them.
I guess the `CLAMP` macro can stay

File: src/code/m_lights.c (hunk 6 above)
Comment (Kelebek1) [reply src/code/m_lights.c:228]: Yes, sizeof returns a u32 and it does divu rather than div without it.

File: src/code/m_lights.c (hunk 7 above)
Comment (Kelebek1) [reply src/code/m_lights.c:12]: Because it's not used outside the file?

Comment (AngheloAlf) [reply src/code/m_lights.c:12]: But this is bss and you know we avoid using `static` on bss symbols when possible

File: src/code/m_lights.c (hunk 11 above)
Comment (Kelebek1) [reply src/code/m_lights.c:47]: *Could be*, do you want me to change it? I left the ones that don't match GC.

File: src/code/m_lights.c [hunk 18]
```diff
@@ -0,0 +1,362 @@
...
+        light->l.dir[0] = lightInfo->diffuse.x;
+        light->l.dir[1] = lightInfo->diffuse.y;
+        light->l.dir[2] = lightInfo->diffuse.z;
+    }
+}
+
+void LightsN_list_check(LightsN* lights, LightNode* node, Vec3f* pos) {
+    static light_point_proc poslight_type_proc[] = { LightsN__point_proc, LightsN__diffuse_proc, LightsN__point_proc };
+    static light_P_point_proc light_type_proc[] = { LightsN__P_point_proc, LightsN__diffuse_proc,
+                                                    LightsN__P_point_proc };
```
Comment (hensldm) [inline src/code/m_lights.c:183]:
```suggestion
    static light_P_point_proc light_type_proc[] = { LightsN__P_point_proc, LightsN__diffuse_proc,
                                                    LightsN__P_point_proc, };
```

Comment (Kelebek1) [reply src/code/m_lights.c:183]: It looked worse doing that, but ok.

**Review by hensldm — APPROVED** [review]

File: linker_scripts/jp/symbol_addrs_code.txt [hunk 19]
```diff
@@ -6087,8 +6087,6 @@ B_FLT_80141DF8_jp = 0x80141DF8; // type:f32
 B_FLT_80141DFC_jp = 0x80141DFC; // type:f32
 B_80141E00_jp = 0x80141E00; //
 B_80141E04_jp = 0x80141E04; // type:s16
-B_80141E10_jp = 0x80141E10; //
-B_80141E18_jp = 0x80141E18; //
```
Comment (AngheloAlf) [inline linker_scripts/jp/symbol_addrs_code.txt:6091]: restore `light_list_buf` and the PR should be ok

**Review by AngheloAlf — APPROVED** [review]

**Pull request merged**